include/rdma/fi_collective.h

```diff
@@ -0,0 +1,294 @@
+ *     without modification, are permitted provided that the following
+ *     conditions are met:
+ *
+ *      - Redistributions of source code must retain the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer.
+ *
+ *      - Redistributions in binary form must reproduce the above
+ *        copyright notice, this list of conditions and the following
+ *        disclaimer in the documentation and/or other materials
+ *        provided with the distribution.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#ifndef FI_COLLECTIVE_H
+#define FI_COLLECTIVE_H
+
+#include <rdma/fi_atomic.h>
+#include <rdma/fi_domain.h>
+#include <rdma/fi_cm.h>
+
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+
+struct fi_ops_av_set {
+	size_t	size;
+	int	(*set_union)(struct fid_av_set *dst,
+			const struct fid_av_set *src);
+	int	(*intersect)(struct fid_av_set *dst,
+			const struct fid_av_set *src);
+	int	(*diff)(struct fid_av_set *dst, const struct fid_av_set *src);
+	int	(*insert)(struct fid_av_set *set, fi_addr_t addr);
+	int	(*remove)(struct fid_av_set *set, fi_addr_t addr);
+};
+
+struct fid_av_set {
+	struct fid		fid;
+	struct fi_ops_av_set	*ops;
+};
+
+struct fi_collective_attr {
+	struct fi_atomic_attr	datatype_attr;
+	size_t			max_members;
+	uint64_t		mode;
+};
+
+#ifdef FABRIC_DIRECT
+#include <rdma/fi_direct_collective.h>
+#endif /* FABRIC_DIRECT */
+
+#ifndef FABRIC_DIRECT_COLLECTIVE
+
+static inline int
+fi_av_set(struct fid_av *av, struct fi_av_set_attr *attr,
+	  struct fid_av_set *av_set, void * context)
+{
+	return FI_CHECK_OP(av->ops, struct fi_ops_av, av_set) ?
+		av->ops->av_set(av, attr, av_set, context) : -FI_ENOSYS;
+}
+
+static inline int
+fi_av_set_union(struct fid_av_set *dst, const struct fid_av_set *src)
+{
+	return dst->ops->set_union(dst, src);
+}
+
+static inline int
+fi_av_set_intersect(struct fid_av_set *dst, const struct fid_av_set *src)
+{
+	return dst->ops->intersect(dst, src);
+}
+
+static inline int
+fi_av_set_diff(struct fid_av_set *dst, const struct fid_av_set *src)
+{
+	return dst->ops->diff(dst, src);
+}
+
+static inline int
+fi_av_set_insert(struct fid_av_set *set, fi_addr_t addr)
+{
+	return set->ops->insert(set, addr);
+}
+
+static inline int
+fi_av_set_remove(struct fid_av_set *set, fi_addr_t addr)
+{
+	return set->ops->remove(set, addr);
+}
+
+static inline int
+fi_join_collective(struct fid_ep *ep, const struct fid_av_set *set,
+		   uint64_t flags, struct fid_mc **mc, void *context)
```

Comment:
I believe MPI will have issues mapping to this API when not using comm world as the base communicator.  We likely need to take as input an fi_addr that references an existing collective communication group.